File: Roslab2/Task3/cpp_first_pkg/src/node2.cpp
#include <chrono>
#include<string>
#include <iostream>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class Node2 : public rclcpp::Node
{
public:
    
    Node2():Node("string_subscriber")
    {
        RCLCPP_INFO(this->get_logger(),"string_subscriber has started");
        string_subscriber_ = this->create_subscription<std_msgs::msg::String>("str_topic",10, std::bind(&Node2::timer_cb_sub, this, std::placeholders::_1));
        flag_publisher_ = this->create_publisher<std_msgs::msg::String>("flag_topic",10);
        timer_ = this->create_wall_timer(std::chrono::milliseconds(500), std::bind(&Node2::timer_cb_pub, this) );
        
    }
private:
    void timer_cb_sub(const std_msgs::msg::String::SharedPtr string_msg)
    {
        RCLCPP_INFO(this->get_logger(),"'%s' ",string_msg->data.c_str());
        if(*(string_msg->data.c_str()+22)=='5'){
           Node2::flag=1;
        }
    }
     void timer_cb_pub()
    {
        std_msgs::msg::String flag_msg = std_msgs::msg::String();
        flag_msg.data = std::to_string(Node2::flag);
        flag_publisher_->publish(flag_msg);
        //RCLCPP_INFO(this->get_logger(),flag_msg.data);
        if (Node2::flag==1){
            Node2::flag=0;
        }
        
        
    }
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr string_subscriber_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr flag_publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    int flag='0';
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc,argv);
    rclcpp::spin(std::make_shared<Node2>());
    rclcpp::shutdown();
    return 0;
}

File: Roslab8/fourth_pkg/fourth_pkg/node2.py
#!/usr/bin/env python3 
import  rclpy
from rclpy.node import Node
from example_interfaces.msg import String

class my_node(Node):
    def __init__(self):
        super().__init__("node2")
        self.counter=0
        self.flag= True
        self.get_logger().info("Node2 is Started")
        self.create_subscription(String,"String_topic",self.timer_call_sub,10)

        

    def timer_call_sub(self,msg_String):
        self.get_logger().info("I Heard: "+ msg_String.data)
        

def main (args=None):
    rclpy.init(args=args)
    node=my_node()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__=="__main__":
    main()

